File: solutions/typescript/game-of-life/1/game-of-life.ts
export class GameOfLife {
  constructor(private matrix: number[][]) {
    if (matrix.some(r => r.length !== matrix[0]?.length ?? 0)) {
      throw new Error('Uneven matrix row length')
    }
    matrix.forEach(r => {
      if (r.some(c => c !== 0 && c !== 1)) {
        throw new Error('Invalid number in matrix')
      }
    })
  }

  public tick(): void {
    const rows = this.matrix.length
    const cols = this.matrix[0]?.length ?? 0
    const toOne = []
    const toZero = []
    for (let i = 0; i < rows; ++i) {
      for (let j = 0; j < cols; ++j) {
        switch (this.aliveNeighbors(i, j)) {
          case 2:
            break
          case 3:
            toOne.push([i, j])
            break
          default:
            toZero.push([i, j])
        }
      }
    }
    toOne.forEach(([i, j]) => this.matrix[i][j] = 1)
    toZero.forEach(([i, j]) => this.matrix[i][j] = 0)
  }

  private aliveNeighbors(i: number, j: number): number {
    const rows = this.matrix.length
    const cols = this.matrix[0]?.length ?? 0
    let acc = 0
    for (let di = -1; di <= 1; ++di) {
      for (let dj = -1; dj <= 1; ++dj) {
        if (di === 0 && dj === 0) {
          continue
        }
        if (0 <= i + di && i + di < rows && 0 <= j + dj && j + dj < cols) {
          acc += this.matrix[i + di][j + dj]
        }
      }
    }
    return acc
  }

  public state(): unknown {
    return structuredClone(this.matrix)
  }
}

File: solutions/lua/game-of-life/1/game-of-life.lua
local function tick(matrix)
  local rows = #matrix
  local cols = matrix[1] and #(matrix[1]) or 0
  local function alive(i, j)
    local total = 0
    for x = math.max(1, i-1), math.min(i+1, rows) do
      for y = math.max(1, j-1), math.min(j+1, cols) do
        if x ~= i or y ~= j then
          total = total + matrix[x][y]
        end
      end
    end
    return total
  end
  local to_one = {}
  local to_zero = {}
  for i, row in ipairs(matrix) do
    for j, cell in ipairs(row) do
      local n = alive(i, j)
      if n == 3 then
        table.insert(to_one, {i, j})
      elseif n ~= 2 then
        table.insert(to_zero, {i, j})
      end     
    end
  end
  for _, o in ipairs(to_one) do
    matrix[o[1]][o[2]] = 1
  end
  for _, z in ipairs(to_zero) do
    matrix[z[1]][z[2]] = 0
  end
  return matrix
end

return { tick = tick }